[RF DOCS] Active Model Basics Documentation [ci-skip] (#51210)
rails/rails: guides/source/active_model_basics.md

* How to use Active Model in your classes.

--------------------------------------------------------------------------------

What is Active Model?
---------------------

To understand Active Model, you need to know a little about [Active
Record](active_record_basics.html). Active Record is an ORM (Object Relational
Mapper) that connects objects whose data requires persistent storage to a
relational database. However, it has functionality that is useful outside of the
ORM, some of these include validations, callbacks, translations, the ability to
create custom attributes, etc.

Some of this functionality was abstracted from Active Record to form Active
Model. Active Model is a library containing various modules that can be used on
plain Ruby objects that require model-like features but are not tied to any
table in a database.

In summary, while Active Record provides an interface for defining models that
correspond to database tables, Active Model provides functionality for building
model-like Ruby classes that don't necessarily need to be backed by a database.
Active Model can be used independently of Active Record.

Some of these modules are explained below.

> ### API

Reviewer: What do you think about moving the section to `ActiveModel::Model` just after API? Those are the two entry points on Active Model, so users usually would use one or the other, and all the other modules are just what compose those two.

Author: ✅

irb> email_contact.name # Attribute Assignment
=> "David"

irb> email_contact.to_model == email_contact # Conversion
=> true

irb> email_contact.model_name.name # Naming
=> "EmailContact"

irb> EmailContact.human_attribute_name("name") # Translation if the locale is set
=> "Name"

irb> email_contact.valid? # Validations
=> true

irb> empty_contact = EmailContact.new
irb> empty_contact.valid?
=> false
```

Any class that includes `ActiveModel::API` can be used with `form_with`,
`render` and any other [Action View helper
methods](https://api.rubyonrails.org/classes/ActionView/Helpers.html), just like
Active Record objects.

For example, `form_with` can be used to create a form for an `EmailContact`
object as follows:

```html+erb
<%= form_with model: EmailContact.new do |form| %>
  <%= form.text_field :name %>
<% end %>
```

which results in the following HTML:

```html
<form action="/email_contacts" method="post">
  <input type="text" name="email_contact[name]" id="email_contact_name">
</form>
```

`render` can be used to render a partial with the object as a local variable:

```html+erb
<%= render "email_contact", email_contact: @email_contact %>
```

> ### Attributes

Reviewer: As this isn't included in `API`, should not this be only mentioned later?

Reviewer 2: IMO, a beginner will likely read this in order, so I'd love if the most common things appeared first in the document.

irb> email_contact.name # Attribute Assignment
=> "David"

irb> email_contact.to_model == email_contact # Conversion
=> true

irb> email_contact.model_name.name # Naming
=> "EmailContact"

irb> EmailContact.human_attribute_name("name") # Translation if the locale is set
=> "Name"

irb> email_contact.valid? # Validations
=> true

irb> empty_contact = EmailContact.new
irb> empty_contact.valid?
=> false
```

Any class that includes `ActiveModel::API` can be used with `form_with`,
`render` and any other [Action View helper
methods](https://api.rubyonrails.org/classes/ActionView/Helpers.html), just like
Active Record objects.

For example, `form_with` can be used to create a form for an `EmailContact`
object as follows:

```html+erb
<%= form_with model: EmailContact.new do |form| %>
  <%= form.text_field :name %>
<% end %>
```

which results in the following HTML:

```html
<form action="/email_contacts" method="post">
  <input type="text" name="email_contact[name]" id="email_contact_name">
</form>
```

`render` can be used to render a partial with the object as a local variable:

```html+erb
<%= render "email_contact", email_contact: @email_contact %>

> ```

Author: Thanks @carlosantoniodasilva, I'm also going to tweak the text a bit so that it's more explicit.

### Model

> [`ActiveModel::Model`](https://api.rubyonrails.org/classes/ActiveModel/Model.html)
> provides the required interface to allow an object to interact with Action Pack
> and Action View. Currently, it only includes [ActiveModel::API](#api) but it
> will be extended in the future to add more functionality.

Reviewer 3: This is how the Api section starts:

![Screenshot 2024-03-06 at 16 51 28](https://github.com/rails/rails/assets/26328/7288ca17-1e4b-4429-b8fa-6aeb34d5d34b)

I am afraid there's not enough distinction between the two since Model here starts very similarly. Here's a small suggestion that might clarify that a bit:

```suggestion
[`ActiveModel::Model`](https://api.rubyonrails.org/classes/ActiveModel/Model.html)
includes [ActiveModel::API](#api) to interact with Action Pack and Action View
by default, and is the recommended approach to implement model-like Ruby
classes. It will be extended in the future to add more functionality.
```

#### Method: `alias_attribute`

`ActiveModel::AttributeMethods` provides aliasing of attribute methods using
`alias_attribute`.

The example below creates an alias attribute for `name` called `full_name`. They
return the same value, but the alias `full_name` better reflects that the
attribute includes a first name and last name.

```ruby
class Person
  include ActiveModel::AttributeMethods

  attribute_method_suffix "_short?"
  define_attribute_methods :name

  attr_accessor :name

  alias_attribute :full_name, :name

  private
    def attribute_short?(attribute)
      public_send(attribute).length < 5
    end
end
```

```irb
irb> person = Person.new
irb> person.name = "Joe Doe"
irb> person.name
=> "Joe Doe"

# `full_name` is the alias for `name`, and returns the same value
irb> person.full_name
=> "Joe Doe"
irb> person.name_short?
=> false

# `full_name_short?` is the alias for `name_short?`, and returns the same value
irb> person.full_name_short?
=> false
```

### Callbacks

[`ActiveModel::Callbacks`](https://api.rubyonrails.org/classes/ActiveModel/Callbacks.html)
gives plain Ruby objects [Active Record style

> callbacks](https://guides.rubyonrails.org/active_record_callbacks.html). The

Reviewer 3: Minor: Let's just link to the guide without the full URL: (so local generated guides point to other local ones, etc.)

```suggestion
callbacks](active_record_callbacks.html). The
```

end

class PersonCallbacks
  def self.before_create(obj)
    # `obj` is the Person instance that the callback is being called on
  end
end
```

#### Aborting Callbacks

The callback chain can be aborted at any point in time by throwing `:abort`.
This is similar to how Active Record callbacks work.

In the example below, since we throw `:abort` before an update in the `reset_me`
method, the remaining callback chain including `before_update` will be aborted,
and the body of the `update` method won't be executed.

```ruby
class Person
  extend ActiveModel::Callbacks

  define_model_callbacks :update

  before_update :reset_me
  after_update :finalize_me
  around_update :log_me

  def update
    run_callbacks(:update) do
      puts "update method called"
    end
  end

  private
    def reset_me
      puts "reset_me method: called before the update method"
      throw :abort
      puts "reset_me method: some code after abort"
    end

    def finalize_me
      puts "finalize_me method: called after the update method"
    end

    def log_me
      puts "log_me method: called around the update method"
      yield
      puts "log_me method: block successfully called"
    end
end
```

> ```

Reviewer 3: <details><summary>For proper highlighting:</summary>
<p>

![Screenshot 2024-03-06 at 17 20 13](https://github.com/rails/rails/assets/26328/17207f7d-0252-4721-a2c5-864a8a69429d)

</p>
</details> 

````suggestion
```irb
````

  `define_attribute_methods`.
3. Call `[attr_name]_will_change!` before each change to the tracked attribute.
4. Call `changes_applied` after the changes are persisted.
5. Call `clear_changes_information` when you want to reset the changes
   information.
6. Call `restore_attributes` when you want to restore previous data.

You can then use the methods provided by `ActiveModel::Dirty` to query the
object for its list of all changed attributes, the original values of the
changed attributes, and the changes made to the attributes.

Let's consider a `Person` class with attributes `first_name` and `last_name` and
determine how we can use `ActiveModel::Dirty` to track changes to these
attributes.

```ruby
class Person
  include ActiveModel::Dirty

  attr_reader :first_name, :last_name
  define_attribute_methods :first_name, :last_name

  def initialize
    @first_name = nil
    @last_name = nil
  end

  def first_name=(value)
    first_name_will_change! unless value == @first_name
    @first_name = value
  end

  def last_name=(value)
    last_name_will_change! unless value == @last_name
    @last_name = value
  end

  def save

>     # Persist data - clears dirty data and moves `changes` to `previously_changed`.

Reviewer 3: I believe this should be `previous_changes`:

```suggestion
    # Persist data - clears dirty data and moves `changes` to `previous_changes`.
```

    changes_applied
  end

  def reload!
    # Clears all dirty data: current changes and previous changes.
    clear_changes_information
  end

  def rollback!

>    # Restores all previous data of the provided attributes.
>     restore_attributes

Reviewer 3: Nit: indent

```suggestion
    # Restores all previous data of the provided attributes.
    restore_attributes
```

```irb
irb> Person.model_name.human
=> "Person"
```

**`collection`** removes the namespace and returns the plural snake_cased name.
It is generally used by Action Pack and/or Action View helpers to aid in
rendering the name of partials/forms.

```irb
irb> Person.model_name.collection
=> "people"
```

**`param_key`** returns a string to use for params names.

```irb
irb> Person.model_name.param_key
=> "person"
```

**`i18n_key`** returns the name of the i18n key. It underscores the model name
and then returns it as a symbol.

```irb
irb> Person.model_name.i18n_key
=> :person
```

**`route_key`** returns a string to use while generating route names.

```irb
irb> Person.model_name.route_key
=> "people"
```

**`singular_route_key`** returns a string to use while generating route names.

```irb
irb> Person.model_name.singular_route_key
=> "person"
```

**`uncountable?`** identifies whether the class name of a record or class is
uncountable.

```irb
irb> Person.model_name.uncountable?
=> false
```

NOTE: Some `Naming` methods, like `param_key`, `route_key` and
`singular_route_key`, differ for namespaced models based on whether it's inside

> an isolated [Engine](https://guides.rubyonrails.org/engines.html).

Reviewer 3: ```suggestion
an isolated [Engine](engines.html).
```

#### Customize the Name of the Model

Sometimes you may want to customize the name of the model that is used in form
helpers and URL generation. This can be useful in situations where you want to
use a more user-friendly name for the model, while still being able to reference
it using its full namespace.

For example, let's say you have a `Person` namespace in your Rails application,
and you want to create a form for a new `Person::Profile`. By default,
Rails would generate the form with the URL `/person/profiles`, which includes
the namespace `person`. However, if you want the URL to simply point to
`profiles` without the namespace, you can customize the `model_name` method like
this:

```ruby
module Person
  class Profile
    include ActiveModel::Model

    def self.model_name
      ActiveModel::Name.new(self, nil, "Profile")
    end
  end
end
```

> With this setup, when you use the `form_for` helper to create a form for

Reviewer 4: ```suggestion
With this setup, when you use the `form_with` helper to create a form for
```

Author: Thank you for catching this!

  def attributes=(hash)
    hash.each do |key, value|
      public_send("#{key}=", value)
    end
  end

  def attributes
    { "name" => nil }
  end
end
```

Now it is possible to create an instance of `Person` and set attributes using
`from_json`.

```irb
irb> json = { name: "Bob" }.to_json
=> "{\"name\":\"Bob\"}"

irb> person = Person.new

irb> person.from_json(json)
=> #<Person:0x00000100c773f0 @name="Bob">

irb> person.name
=> "Bob"
```

### Translation

[`ActiveModel::Translation`](https://api.rubyonrails.org/classes/ActiveModel/Translation.html)
provides integration between your object and the [Rails internationalization

> (i18n) framework](https://guides.rubyonrails.org/i18n.html).

Reviewer 3: ```suggestion
(i18n) framework](i18n.html).
```

```

With the `human_attribute_name` method, you can transform attribute names into a
more human-readable format. The human-readable format is defined in your locale
file(s).

```yaml
# config/locales/app.pt-BR.yml
pt-BR:
  activemodel:
    attributes:
      person:
        name: "Nome"
```

```irb
irb> Person.human_attribute_name("name")
=> "Name"

irb> I18n.locale = :"pt-BR"
=> :"pt-BR"
irb> Person.human_attribute_name("name")
=> "Nome"
```

### Validations

[`ActiveModel::Validations`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html)
adds the ability to validate objects and it is important for ensuring data
integrity and consistency within your application. By incorporating validations
into your models, you can define rules that govern the correctness of attribute
values, and prevent invalid data.

```ruby
class Person
  include ActiveModel::Validations

  attr_accessor :name, :email, :token

  validates :name, presence: true

>   validates_format_of :email, with: /\A([^\s]+)((?:[-a-z0-9]\.)[a-z]{2,})\z/i

Reviewer 3: I believe the idea was to show the different "variations", but the case that `validates` is generally preferred, makes for a good point on using the former and not documenting the latter in the guides. We can change it. :+1:

- [`validates_with`](https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates_with):
  Passes the record off to the class or classes specified and allows them to add
  errors based on more complex conditions.

- [`validates_each`](https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates_each):
  Validates each attribute against a block.

Some of the options below can be used with certain validators. To determine if
the option you're using can be used with a specific validator, read through the
documentation [here](https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html).

- `:on`: Specifies the context in which to add the validation. You can pass a
  symbol or an array of symbols. (e.g. `on: :create` or `on:
  :custom_validation_context` or `on: [:create, :custom_validation_context]`).
  Validations without an `:on` option will run no matter the context. Validations
  with some `:on` option will only run in the specified context. You can pass the
  context when validating via `valid?(:context)`.

- `:if`: Specifies a method, proc or string to call to determine if the

>   validation should occur (e.g. `if: :allow_validation`, or `if: Proc.new {

Author: I made some tweaks here: In this `if` example,  I've updated it to use a lambda to show that you can use a simpler form. In the `unless` I've kept it as a Proc to indicate a variation.
